src/main/java/com/terracottatech/frs/SecurePutAction.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright IBM Corp. 2024, 2025
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.terracottatech.frs;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+import com.terracottatech.frs.action.ActionCodec;
+import com.terracottatech.frs.action.ActionFactory;
+import com.terracottatech.frs.compaction.Compactor;
+import com.terracottatech.frs.object.ObjectManager;
+import com.terracottatech.frs.util.ByteBufferUtils;
+
+public class SecurePutAction extends PutAction {
+  /*
+   * CipheredPutAction Header
+   * 4 bytes - idByteCount
+   * 4 bytes - keyByteCount
+   * 4 bytes - ivByteCount
+   * 4 bytes - valueByteCount
+   * 8 bytes - invalidatedLsn
+   */
+  public static final long CIPHERED_PUT_ACTION_OVERHEAD = 24L;
+
+  @SuppressWarnings("rawtypes")
+  public static final ActionFactory<ByteBuffer, ByteBuffer, ByteBuffer> FACTORY = (
+      ObjectManager<ByteBuffer, ByteBuffer, ByteBuffer> objectManager, ActionCodec codec, ByteBuffer[] buffers) -> {
+    int idLength = ByteBufferUtils.getInt(buffers);
+    int keyLength = ByteBufferUtils.getInt(buffers);
+    int ivLength = ByteBufferUtils.getInt(buffers);
+    int valueLength = ByteBufferUtils.getInt(buffers);
+    long invalidatedLsn = ByteBufferUtils.getLong(buffers);
+    ByteBuffer id = ByteBufferUtils.getBytes(idLength, buffers);
+    ByteBuffer key = ByteBufferUtils.getBytes(keyLength, buffers);
+    ByteBuffer iv = ByteBufferUtils.getBytes(ivLength, buffers);
+    CipherManager cipherManager = codec.getCipherManager();
```

Comment:
It seems wrong to retrieve this every time. It also seems wrong that the codec exposes this method... but that might be a harder fix. Thinking you could move the two encrypted action registrations to a common `EncryptionActions` class then you can retrieve the `CipherManager` in a single place and pass it through to the two factories to be stored as an instance variable?

Reply:
Will take into account.